File: gpio4j/src/main/java/org/overworld/gpio4j/serial/PinType.java
package org.overworld.gpio4j.serial;

import java.util.Optional;
import java.util.stream.Stream;

import lombok.Getter;

public enum PinType {
	DIGITAL_IN(1), DIGITAL_IN_PULLUP(2), DIGITAL_OUT(3), ANALOG_IN(4), ANALOG_OUT(5), COSINE(6), BUFFER(7), PWM(8);

	@Getter
	private int id;

	private PinType(int type) {
		this.id = type;
	}
	
	public static Optional<PinType> forId(int id) {
		return Stream.of(PinType.values()).filter(pt -> id == pt.getId()).findFirst();
	}
}

File: BoardDriver/src/PinType.cpp
#include "PinType.h"

class PinTypeHelper {
public:
    static uint8_t getId(PinType type) {
        return static_cast<uint8_t>(type);
    }

    static PinType forId(uint8_t id) { // @suppress("No return")
        static const std::vector<PinType> allTypes = {
            PinType::DIGITAL_IN,
            PinType::DIGITAL_IN_PULLUP,
            PinType::DIGITAL_OUT,
            PinType::ANALOG_IN,
            PinType::ANALOG_OUT,
            PinType::COSINE,
            PinType::BUFFER,
            PinType::PWM
        };

        auto it = std::find_if(allTypes.begin(), allTypes.end(),
                               [id](PinType pt) { return getId(pt) == id; });

        if (it != allTypes.end()) {
            return *it;
        }
    }
};
